src/vm/docs/mod.rs

```diff
@@ -54,10 +54,18 @@ const TX_SENDER_KEYWORD: KeywordAPI = KeywordAPI {
     name: "tx-sender",
     output_type: "principal",
     description: "Returns the current context's transaction sender. This is either the principal that signed and submitted the transaction,
-or if `as-contract` was used to change the tx-sender context, it will be that contract's principal.",
+or if `contract-call!` was used to change the tx-sender context, it will be that contract's principal.",
     example: "(print tx-sender) ;; Will print out a Stacks address of the transaction sender",
 };
 
+const TX_ORIGIN_KEYWORD: KeywordAPI = KeywordAPI {
+    name: "tx-origin",
+    output_type: "principal",
+    description: "Returns the original sender of the current transaction, or if `as-contract` was called to modify the sending context, it returns that
```

Comment:
I think I'm lost.
> Returns the original sender of the current transaction

👍

> if `as-contract` was called to modify the sending context, it returns that contract principal.

Isn't it the point of introducing `tx-sender`?

Let's say Alice is performing a transaction handled by Contract X, calling Contract Y, calling Contract Z.

| stack  | tx-origin  | tx-sender |
|:------------- |:---------------:| -------------:|
| X      | Alice |         Alice |

| stack  | tx-origin  | tx-sender |
|:------------- |:---------------:| -------------:|
| Y      | Alice      |         X|
| X      | Alice     |         Alice |


| stack  | tx-origin  | tx-sender |
|:------------- |:---------------:| -------------:|
| Z | Alice        |          Y |
| Y      | Alice        |         X|
| X      | Alice |         Alice |


If Contract X and Contract Y have the ability to prefix their `contract-call!` with `as-contract!`, we could end up with that:

| stack  | tx-origin  | tx-sender |
|:------------- |:---------------:| -------------:|
| Z      | Y        |          Y |
| Y      | X        |         X|
| X      | Alice |         Alice |


Shouldn't we get rid of `as-contract!` instead?

Reply:
Right. I'm not sure, and thought about this a bit during this PR. As is, this allows a contract to mimic the sending of a transaction. Otherwise, a contract principal _could never_ be the `tx-origin`. I'm not sure which is the right approach. I think, perhaps a better option is to leave `tx-origin` out and `tx-sender` alone entirely for now (keeping the current behavior in `develop`), and start working on first-class assets. Once we see how the first-class assets can be used, then we can revisit this policy.

Reply:
Okay -- I talk about this in #1051 -- I think the right approach is described there and implemented now in this PR.

We have `tx-sender` and `contract-caller` functioning as you describe, except that `as-contract` changes _both_ of them at the same time. This allows a contract to _simulate_ a transaction send (as the contract principal), while also allowing contracts to check that a given call was a "direct call", by performing a `(eq? tx-sender contract-caller)` test. We'll want to update SIP-001 with a description of that, as well as any documentation introducing the concept of principals and checking for permissions. Finally, introducing first class asset types should alleviate the burden of implementing this correctly for most contract authors.